### PR TITLE
Change grpc address formet

### PR DESCRIPTION
**Pull request by mrwwheat7**

### CODE REVIEW

File: src/client.rs [hunk 1]
```diff
@@ -50,11 +51,34 @@ impl Client {
...
+        if cfg.endpoints.len() > 1 {
+            let socket_address: SocketAddr = cfg
+                .endpoints
+                .first()
+                .unwrap_or_else(|| panic!("Fail to get the first endpoint"))
+                .parse()
+                .unwrap_or_else(|e| {
+                    panic!(
+                        "Fail to parse enpoint to socket address, the error is {}",
+                        e
```
Comment (pwang7) [inline src/client.rs]: add comma

File: src/client.rs [hunk 2]
```diff
@@ -50,11 +51,34 @@ impl Client {
...
-        let end_points = cfg.endpoints.join(",");
+        if cfg.endpoints.is_empty() {
+            panic!("Empty etcd endpoints");
+        }
+        let mut end_points = cfg.endpoints.join(",");
         let env = Arc::new(EnvBuilder::new().build());
+        if cfg.endpoints.len() > 1 {
+            let socket_address: SocketAddr = cfg
+                .endpoints
+                .first()
```
Comment (pwang7) [inline src/client.rs:62]: only accept first etcd address?

**Review by pwang7 — CHANGES_REQUESTED** [review]

File: src/client.rs (hunk 2 above)
Comment (mrwwheat7) [reply src/client.rs:62]: No, it will only use the first address to check the schema, not only accept first etcd address

File: src/client.rs [hunk 3]
```diff
@@ -50,14 +52,71 @@ impl Client {
...
         // }
         // Ok(Channel::balance_list(endpoints.into_iter()))
-        let end_points = cfg.endpoints.join(",");
+        if cfg.endpoints.is_empty() {
+            panic!("Empty etcd endpoints");
+        }
+        let mut end_points = cfg.endpoints.join(",");
         let env = Arc::new(EnvBuilder::new().build());
+        if cfg.endpoints.len() > 1 {
+            for i in 0..cfg.endpoints.len().overflow_sub(1) {
```
Comment (pwang7) [inline src/client.rs]: use functional style to loop, instead of using for loop

File: src/client.rs [hunk 4]
```diff
@@ -50,14 +52,71 @@ impl Client {
...
+        }
         let ch = ChannelBuilder::new(env)
             .load_balancing_policy(LbPolicy::RoundRobin)
-            .connect(&format!("{}:{}", "ipv4", end_points));
+            .connect(end_points.as_str());
         Ok(ch)
     }
 
+    /// Checks if two ip address has same schema.
+    fn is_same_ip_address_schema(ip1: &str, ip2: &str) -> bool {
```
Comment (pwang7) [inline src/client.rs]: this function can be a closure when using functional style loop

**Review by pwang7 — CHANGES_REQUESTED** [review]

File: src/client.rs [hunk 5]
```diff
@@ -50,14 +51,60 @@ impl Client {
...
+        }
         let ch = ChannelBuilder::new(env)
             .load_balancing_policy(LbPolicy::RoundRobin)
-            .connect(&format!("{}:{}", "ipv4", end_points));
+            .connect(end_points.as_str());
         Ok(ch)
     }
 
+    /// Checks if two ip address has same schema.
+    fn is_same_ip_address_schema(ip1: SocketAddr, ip2: SocketAddr) -> bool {
```
Comment (pwang7) [inline src/client.rs]: should be a closure, instead of a function

File: src/client.rs [hunk 6]
```diff
@@ -50,14 +51,60 @@ impl Client {
...
+                    panic!("Fail to get the first endpoint")
+                }
+            });
+
+            if socket_address.is_ipv4() {
+                end_points = format!("{}:{}", "ipv4", end_points)
+            } else if socket_address.is_ipv6() {
+                end_points = format!("{}:{}", "ipv6", end_points)
+            } else {
+                panic!("Unsupport etcd endpoint address");
```
Comment (pwang7) [inline src/client.rs]: panic!("unsupported etcd address: {}", socket_address)

File: src/client.rs [hunk 7]
```diff
@@ -50,14 +51,60 @@ impl Client {
...
+                            e,
+                        )
+                    }),
+                );
+                if !is_same_schema {
+                    panic!("Fail to get the first endpoint")
+                }
+            });
+
+            if socket_address.is_ipv4() {
```
Comment (pwang7) [inline src/client.rs:85]: check address type before for each loop

**Review by pwang7 — CHANGES_REQUESTED** [review]

File: src/client.rs (hunk 7 above)
Comment (mrwwheat7) [reply src/client.rs:85]: The for each loop is used to check the endpoints has same type of ip address schema, if it is failed, then it just panic. So we should not check address type before for each loop

**Review by pwang7 — APPROVED** [review]

**Pull request merged**